pykafka/producer.py

```diff
@@ -535,6 +534,8 @@ def queue_reader():
                 except Exception:
                     # surface all exceptions to the main thread
                     self.producer._worker_exception = sys.exc_info()
+                    if not self.slot_available.is_set():
+                        self.slot_available.set()
```

Comment:
I think I'd be ok with moving this logic into a `OwnedBroker.cleanup()` function that is called not in the exception handler, but after the `while self.running` loop. This will make it clearer why we're messing with `slot_available` here and avoid potential confusion when we revisit this code. 
```python
def cleanup(self):
    if not self.slot_available.is_set():
        self.slot_available.set()
```
and
```python
while self.running:
    ....
    except Exception:
        ...
self.cleanup()
```
If you add those two things to this pull request, it should be fine to merge.